sql: support COMMENT ON SCHEMA

ekalinin opened the pull request:
Fixes #67689 

This change adds support for SCHEMA comment.

Release note (sql change): This change adds associating
comment to SQL schema using PostgreSQL's 
`COMMENT ON SCHEMA` syntax.

ekalinin committed:
sql: support COMMENT ON SCHEMA

This change adds support for SCHEMA commenting.

Release note (sql change): This change adds associating
comment to SQL schema using PostgreSQL's COMMENT ON SCHEMA syntax.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68606)
<!-- Reviewable:end -->

knz reviewed (commented):
This PR is missing the necessary changes to `sql/pg_catalog.go` to generate the values in `pg_catalog.pg_description`. This will also need a test about how to retrieve the schema comment via `pg_catalog.obj_description()`.

You'll probably need input from someone in the SQL schema team about how to populate pg_description properly.

> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-:-MgoylVr6jysV-lGfQu6:bdlorgr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin)

___
*[pkg/sql/drop_schema.go, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-Mgox_9o74TAAlnFXGFe:-Mgox_9o74TAAlnFXGFf:b-8wbta1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9008f0807135eb328f811a6628b907cd12821f2e/pkg/sql/drop_schema.go#L185)):*
> ```Go
> 		}
> 
> 		if err := p.removeSchemaComment(params.ctx, sc.GetID()); err != nil {
> ```

This needs to go to `dropSchemaImpl` otherwise the comment is not dropped upon DROP DATABASE CASCADE


<!-- Sent from Reviewable.io -->

ekalinin commented:
> This PR is missing the necessary changes to `sql/pg_catalog.go` to generate the values in `pg_catalog.pg_description`. This will also need a test about how to retrieve the schema comment via `pg_catalog.obj_description()`.
> 
> You'll probably need input from someone in the SQL schema team about how to populate pg_description properly.
>

Thanks. Added.

> This needs to go to dropSchemaImpl otherwise the comment is not dropped upon DROP DATABASE CASCADE

Done.

knz reviewed (commented):
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin)

___
*[pkg/sql/pg_catalog.go, line 1398 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkT9i2JaHPUeTuiSL:-MgqkT9i2JaHPUeTuiSM:b-mmvxyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/pg_catalog.go#L1398)):*
> ```Go
> 			case keys.SchemaCommentType:
> 				objID = tree.NewDOid(tree.MustBeDInt(objID))
> 				classOid = tree.NewDOid(catconstants.PgCatalogSchemaID)
> ```

We'll need a review from SQL Schema for this change.
(@ajwerner?)
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MgqkgGH-Rwkp3gtO3wX:b-vdsegf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
> ```
> ----
> objoid  classoid    objsubid  description
> 94      4294967022  0         testing schema
> ```

I think we need to test this via `pg_catalog.obj_description()` because I don't think we wish to assert particular object IDs inside unit tests.

Someone from SQL experience would know better.
(@rafiss?)


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/sem/builtins/pg_builtins.go:
```diff
@@ -1942,6 +1942,8 @@ func getCatalogOidForComments(catalogName string) (id int, ok bool) {
 	case "pg_database":
 		return catconstants.PgCatalogDatabaseTableID, true
+	case "pg_schema":
+		return catconstants.PgCatalogSchemaID, true
```
I checked the postgres documentation and I couldn't find any mention of `pg_schema` in there. Strictly speaking the table for schemas in the pg catalog is `pg_namespace`. As far as comments are concerned, it seems they might actually belong in `pg_shdescription` or `pg_description`. Perhaps @rafiss can enlighten us?

knz replied:
a quick test inside `psql` tells me that schema comments get stored inside `pg_description`.

ekalinin commented:
> I think we need to test this via pg_catalog.obj_description() because I don't think we wish to assert particular object IDs inside unit tests.

Fixed.

> I checked the postgres documentation and I couldn't find any mention of pg_schema in there. Strictly speaking the table for schemas in the pg catalog is pg_namespace

Fixed.

otan reviewed (commented):
> Reviewed 13 of 14 files at r1, 2 of 6 files at r2, 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MhGus0VC1LHoXxfRZty:b282zv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think we need to test this via `pg_catalog.obj_description()` because I don't think we wish to assert particular object IDs inside unit tests.

Someone from SQL experience would know better.
(@rafiss?)
</blockquote></details>

i'd probably omit the oids / subids simply because they are potentially unstable.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvFV98nv8A46mYcnL:-MhGvFV98nv8A46mYcnM:b-um91z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5082)):*
> ```
> SELECT obj_description(objoid)
>   FROM pg_catalog.pg_description
>  WHERE description = 'testing schema';
> ```

nit: drop the trailing `;` here and above
___
*[pkg/sql/logictest/testdata/logic_test/schema, line 881 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvI9B97OtFUBj-9oI:-MhGvI9B97OtFUBj-9oJ:b2f5ohj) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/schema#L881)):*
> ```
> 
> statement ok;
> USE comment_db;
> ```

nit: drop the trailing `;` here 
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqaSox7ndfLf79FuDJ-r2-1946:-MhGvUzB5Vr6PQ-l7aXI:b-ppk6vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/sem/builtins/pg_builtins.go#L1946)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

a quick test inside `psql` tells me that schema comments get stored inside `pg_description`.
</blockquote></details>

`pg_description` is correct


<!-- Sent from Reviewable.io -->

ekalinin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MgvhatA2zWy-YJTmHbP:b-qoxx8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i'd probably omit the oids / subids simply because they are potentially unstable.
</blockquote></details>

Should I remove test via `obj_description`?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MhISs1n0Oaj7J_D_umI:bqv3vp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, ekalinin (Eugene Kalinin) wrote…</i></summary><blockquote>

Should I remove test via `obj_description`?
</blockquote></details>

Is there a test for it? We should keep it for sure if there is and add it if it isn't. 


<!-- Sent from Reviewable.io -->

ekalinin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, @otan, and @rafiss)

___
*[pkg/sql/drop_schema.go, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-Mgox_9o74TAAlnFXGFe:-MhKBmKG9ZmdDc3JwuFl:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9008f0807135eb328f811a6628b907cd12821f2e/pkg/sql/drop_schema.go#L185)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs to go to `dropSchemaImpl` otherwise the comment is not dropped upon DROP DATABASE CASCADE
</blockquote></details>

Fixed.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MhKBD0OAgVka_aSHyeu:bis8y3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Is there a test for it? We should keep it for sure if there is and add it if it isn't. 
</blockquote></details>

Yes, there's. Please, see the latest version of the PR.

https://github.com/cockroachdb/cockroach/pull/68606/files#diff-44361d449f2d21bb6beb453a14b8475fe98e73a04bf990d251ed99f055001eb8R5080
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvFV98nv8A46mYcnL:-MhKBMYSCxBkbPpeGuoF:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5082)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: drop the trailing `;` here and above
</blockquote></details>

Fixed.
___
*[pkg/sql/logictest/testdata/logic_test/schema, line 881 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvI9B97OtFUBj-9oI:-MhKBdy59HbItxUPbdH4:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/schema#L881)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: drop the trailing `;` here 
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

ekalinin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, @otan, and @rafiss)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqaSox7ndfLf79FuDJ-r2-1946:-MhKDeBE3LtzfOTA-3OI:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/sem/builtins/pg_builtins.go#L1946)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

`pg_description` is correct
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

otan reviewed (approved):
looks good to me! thanks

bors r+
